File: Example/Example/ViewController.m
//
//  ViewController.m
//  Example
//
//  Created by Jonathan Tribouharet
//

#import "ViewController.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    
    self.scrollView.effect = JT3DScrollViewEffectCards;
    
    self.scrollView.delegate = self; // Use only for animate nextButton and previousButton
    
    [self createCardWithColor];
    [self createCardWithColor];
    [self createCardWithColor];
    [self createCardWithColor];
    
    self.nextButton.backgroundColor = [UIColor colorWithRed:33/255. green:158/255. blue:238/255. alpha:1.];
    self.nextButton.layer.cornerRadius = 5.;
    self.previousButton.layer.cornerRadius = 5.;
}

- (void)createCardWithColor
{
    CGFloat width = CGRectGetWidth(self.scrollView.frame);
    CGFloat height = CGRectGetHeight(self.scrollView.frame);
 
    CGFloat x = self.scrollView.subviews.count * width;
    
    UIView *view = [[UIView alloc] initWithFrame:CGRectMake(x, 0, width, height)];
    view.backgroundColor = [UIColor colorWithRed:33/255. green:158/255. blue:238/255. alpha:1.];
    
    view.layer.cornerRadius = 8.;
    
    [self.scrollView addSubview:view];
    self.scrollView.contentSize = CGSizeMake(x + width, height);
}

#pragma mark - SegmentControl

- (IBAction)didChangeMode:(UISegmentedControl *)sender
{
    JT3DScrollViewEffect effect;
    switch (sender.selectedSegmentIndex) {
        case 0:
            effect = JT3DScrollViewEffectCards;
            break;
        case 1:
            effect = JT3DScrollViewEffectCarousel;
            break;
        case 2:
            effect = JT3DScrollViewEffectDepth;
            break;
        case 3:
            effect = JT3DScrollViewEffectTranslation;
            break;
            
        default:
            break;
    }
    
    self.scrollView.effect = effect;
}

#pragma mark - Next / Previous buttons

- (IBAction)loadNextPage:(id)sender
{
    [self.scrollView loadNextPage:YES];
}

- (IBAction)loadPreviousPage:(id)sender
{
    [self.scrollView loadPreviousPage:YES];
}

- (void)scrollViewDidEndScrollingAnimation:(UIScrollView *)scrollView
{
    [self updateButtons];
}

- (void)scrollViewDidEndDecelerating:(UIScrollView *)scrollView
{
    [self updateButtons];
}

- (void)updateButtons
{
    if(self.scrollView.currentPage >= self.scrollView.subviews.count - 1){
        [self showPreviousButton];
    }
    else{
        [self showNextButton];
    }
}

- (void)showNextButton
{
    [UIView animateWithDuration:.3
                     animations:^{
                         self.leftNextButtonConstraint.constant = 40;
                         [self.view layoutIfNeeded];
                     }];
}

- (void)showPreviousButton
{
    [UIView animateWithDuration:.3
                     animations:^{
                         self.leftNextButtonConstraint.constant = - CGRectGetWidth(self.view.frame);
                         [self.view layoutIfNeeded];
                     }];
}

@end

File: Example/Example/ViewController.swift
//
//  ViewController.swift
//  Example
//
//  Created by Jonathan on 07/06/2017.
//  Copyright © 2017 Jonathan Tribouharet. All rights reserved.
//

import UIKit
import JT3DScrollView

class ViewController: UIViewController, UIScrollViewDelegate {
    
    @IBOutlet weak var scrollView: JT3DScrollView?
    @IBOutlet weak var nextButton: UIButton?
    @IBOutlet weak var previousButton: UIButton?
    
    @IBOutlet weak var leftNextButtonConstraint: NSLayoutConstraint?
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        scrollView?.effect = .cards
        
        scrollView?.delegate = self // Use only for animate nextButton and previousButton
        
        createCardWithColor()
        createCardWithColor()
        createCardWithColor()
        createCardWithColor()
        
        nextButton?.backgroundColor = UIColor(colorLiteralRed: 33/255, green: 158/255, blue: 238/255, alpha: 1)
        nextButton?.layer.cornerRadius = 5
        previousButton?.layer.cornerRadius = 5
    }
    
    func createCardWithColor () {
        guard let scrollView = scrollView else {
            return
        }
        
        let width = scrollView.frame.width
        let height = scrollView.frame.height
        
        let x = CGFloat(scrollView.subviews.count) * width
        
        let view = UIView(frame: CGRect(x: x, y: 0, width: width, height: height))
        view.backgroundColor = UIColor(colorLiteralRed: 33/255, green: 158/255, blue: 238/255, alpha: 1)
        view.layer.cornerRadius = 8
        scrollView.addSubview(view)
        scrollView.contentSize = CGSize(width: x + width, height: height)
    }
    
    //MARK - SegmentControl
    
    @IBAction func didChangeMode(_ sender: UISegmentedControl) {
    
        let effect: JT3DScrollViewEffect
        
        switch (sender.selectedSegmentIndex) {
        case 0:
            effect = .cards
        case 1:
            effect = .carousel
        case 2:
            effect = .depth
        case 3:
            effect = .translation
        default:
            effect = .none
        }
        
        scrollView?.effect = effect
    }
    
    
    // MARK - Next / Previous buttons
    
    @IBAction func loadNextPage(_ sender: UIButton) {
        scrollView?.loadNextPage(animated: true)
    }
    
    @IBAction func loadPreviousPage(_ sender: UIButton) {
        scrollView?.loadPreviousPage(animated: true)
    }
    
    func scrollViewDidEndScrollingAnimation(_ scrollView: UIScrollView) {
        updateButtons()
    }
    
    func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
        updateButtons()
    }
    
    func updateButtons () {
        guard let scrollView = scrollView else {
            return
        }
        
        if(scrollView.currentPage >= scrollView.subviews.count - 1){
            showPreviousButton()
        }
        else{
            showNextButton()
        }
    }
    
    func showPreviousButton () {
        UIView.animate(withDuration: 0.3, animations: { 
            self.leftNextButtonConstraint?.constant = -self.view.frame.width
            self.view.layoutIfNeeded()
        })
    }
    
    func showNextButton () {
        UIView.animate(withDuration: 0.3, animations: {
            self.leftNextButtonConstraint?.constant = 40
            self.view.layoutIfNeeded()
        })
    }

}
